lib/webdriveragent.js

```diff
@@ -46,123 +48,210 @@ class WebDriverAgent {
       'WebDriverAgentRunner',
+      '-sdk',
+      realDevice ? 'iphoneos' : 'iphonesimulator',
       '-destination',
-      `id=${this.sim.udid}`,
-      'test'
+      `platform=${realDevice ? 'iOS' : 'iOS Simulator'},id=${this.device.udid}`,
+      'test',
+      `CODE_SIGN_IDENTITY=${realDevice ? 'iPhone Developer' : '""'}`,
+      `CODE_SIGNING_REQUIRED=${realDevice ? 'YES' : 'NO'}`
     ];
-    return new SubProcess('xcodebuild', args, {cwd: this.bootstrapPath});
+    let xcodebuild = new SubProcess('xcodebuild', args, {cwd: this.bootstrapPath});
+
+    this.setupXcodeLogging(xcodebuild, reject);
+
+    return xcodebuild;
+  }
+
+  setupXcodeLogging (xcodebuild, reject) {
+    xcodebuild.on('output', (d, e) => {
+      // we want to pull out the log file that is created, and highlight it
+      // for diagnostic purposes
+      if (e.indexOf('Writing diagnostic log for test session to') !== -1) {
+        // pull out the first line that begins with the path separator
+        // which should be the line indicating the log file generated
+        this.logLocation = _.first(_.remove(e.trim().split('\n'), (v) => v.indexOf(path.sep) === 0));
+        log.debug(`Log file for xcodebuild test: ${this.logLocation}`);
+      }
+      // log output of xcodebuild, for debugging
+      // log.info(d, e);
+    });
+    xcodebuild.on('exit', (code, signal) => {
+      if (!code && signal === 'SIGTERM') {
+        log.info(`xcodebuild exited with code '${code}' and signal '${signal}'`);
+      }
+      if (!signal && code !== 0) {
+        let err = new Error(`xcodebuild failed with code ${code}`);
+        if (reject) {
+          reject(err);
+          return;
+        } else {
+          throw err;
+        }
+      }
+      this.quit();
+    });
   }
 
   createSimLogsSubProcess () {
     let args = [
       '-f',
       '-n',
       '0',
-      path.resolve(this.sim.getLogDir(), 'system.log')
+      path.resolve(this.device.getLogDir(), 'system.log')
     ];
 
-    return new SubProcess('tail', args);
+    let logs = new SubProcess('tail', args);
+    this.setupLogging(logs, 'Sim');
+    logs.on('exit', (code) => {
+      if (code) {
+        log.errorAndThrow(new Error(`Tailing of simulator log exited with code '${code}'`));
+      }
+      log.debug('Tailing of simulator log exited.');
+    });
+
+    return logs;
+  }
+
+  createRealDeviceLogsSubProcess () {
+    let logs = new SubProcess('idevicesyslog', ['-u', this.device.udid]);
+    this.setupLogging(logs, 'Device');
+    logs.on('exit', (code) => {
+      if (code) {
+        log.errorAndThrow(new Error(`System log exited with code '${code}'`));
+      }
+      log.debug('System log exited');
+    });
+
+    return logs;
+  }
+
+  setupLogging (logs, prefix) {
+    function isPertinentLogLine (line) {
+      return line.length &&
+            (line.indexOf(AGENT_LOG_PREFIX) !== -1 ||
+             line.indexOf(AGENT_RUNNER_LOG_PREFIX) !== -1);
+    }
+    logs.on('output', (d, e) => {
```

Comment:
are `d` and `e` stdout and stderr? if so, (a) these variables are not necessarily that transparent, and (b) i'm pretty sure you never get both at the same time, so you might be able to compress this:

```js
let line = d || e;
if (isPertinentLogLine(line)) {
  agentLog.debug(`${prefix}: ${line.trim()}`);
}
```